src/Appwrite/Platform/Tasks/Backup.php

```diff
@@ -0,0 +1,318 @@
+            } catch (Exception $e) {
+                //todo: send alerts to admin!
+                // todo: Do we want to terminate the script? or wait for next backup iteration?
+                Console::error(date('Y-m-d H:i:s') . ' Error: ' . $e->getMessage());
+            }
+        }, self::BACKUP_INTERVAL_SECONDS);
+    }
+
+    public function start(): void
+    {
+        $start = microtime(true);
+        $time = date('Y_m_d_H_i_s');
+
+        self::log('--- Backup Start ' . $time . ' --- ');
+
+        $filename = $time . '.tar.gz';
+        $local = new Local(self::BACKUPS_PATH . '/' . $this->database . '/full/' . $time);
+        $local->setTransferChunkSize(self::UPLOAD_CHUNK_SIZE);
+
+        $backups = $local->getRoot() . '/files';
+        $tarFile = $local->getPath($filename);
+
+        $this->fullBackup($backups);
+        $this->tar($backups, $tarFile);
+        $this->upload($tarFile, $local);
+
+        self::log('--- Backup Finish ' . (microtime(true) - $start) . ' seconds --- '   . PHP_EOL . PHP_EOL);
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function fullBackup(string $target)
+    {
+        $start = microtime(true);
+        self::log('Xtrabackup start');
+
+        if (!file_exists(self::BACKUPS_PATH)) {
+            throw new Exception('Mount directory does not exist');
+        }
+
+        if (!file_exists($target) && !mkdir($target, 0755, true)) {
+            throw new Exception('Error creating directory: ' . $target);
+        }
+
+        $filename = basename($target);
+        $logfile = $target . '/../backup.log';
+
+        $args = [
+            'xtrabackup',
+            '--user=' . $this->dsn->getUser(),
+            '--password=' . $this->dsn->getPassword(),
+            '--host=' . $this->dsn->getHost(),
+            '--port=' . $this->dsn->getPort(),
+            '--backup',
+            '--strict',
+            '--history="' . $this->database . '|' . pathinfo($filename, PATHINFO_FILENAME) . '"', // PERCONA_SCHEMA.xtrabackup_history
+            '--slave-info',
+            '--safe-slave-backup',
+            '--safe-slave-backup-timeout=300',
+            '--check-privileges', // checks if Percona XtraBackup has all the required privileges.
+            '--target-dir=' . $target,
+            '--compress=' . self::COMPRESS_ALGORITHM,
+            '--compress-threads=' . $this->processors, // Better not using all processors
+            '--parallel=' . $this->processors,
+            '--rsync', // https://docs.percona.com/percona-xtrabackup/8.0/accelerate-backup-process.html
+            '2> ' . $logfile,
+        ];
+
+        $cmd = 'docker exec ' . $this->xtrabackupContainerId . ' ' . implode(' ', $args);
+        shell_exec($cmd);
+
+        $stderr = shell_exec('tail -1 ' . $logfile);
+        if (!str_contains($stderr, 'completed OK!')) {
+            shell_exec('rm -rf ' . $target);
+            throw new Exception(' Backup failed: ' . $stderr);
```

Comment:
Let's add a retry loop here with 5 max attempts before throwing

Reply:
Done!
Hmm 5 times I think is too much, we do not know how long this will take in production, we want space between iterations.